### PR TITLE
Improve ITs performance

### LINKED ISSUE #736: Enable running the regression tests for just one analyzer
To improve the feedback time from the regression tests while developing a new rule we can enable the script to limit the analysis to just that rule. This way we will be able to run the tests in much shorter time, which will enable us to add more projects to the test suite.

### PR DESCRIPTION
**Pull request by michalb-sonar**
Fix #736 
regression-test.ps1 has a new optional parameter, ruleId. Sample use:
`regression-test.ps1 -ruleId S3242`

This change makes tests for a single rule to run in just over 2 minutes. Summary of changes:
1. its/SingleRule.ruleset.template is a template that disables ALL rules. The script then enables only the one we passed as a parameter.
2. Some rules could not be disabled (they had `WellKnownDiagnosticTags.NotConfigurable` attribute). I've made it possible to disable them, in debug builds only.
3. It is not necessary to create a git repository to diff two folders with `git diff`. By not doing it, we save about 40s - so as a side benefit of the change, running all tests will also be quicker.
4. When diffing results for one rule, there is no point of diffing everything - only the reports from the rule we've just run.

### CODE REVIEW

File: sonaranalyzer-dotnet/its/regression-test.ps1 [hunk 1]
```diff
@@ -25,33 +35,121 @@ function Build-Project([string]$ProjectName, [string]$SolutionRelativePath) {
...
-    Copy-Item .\expected -Destination .\actual -Recurse
-    Invoke-InLocation "actual" {
-        Exec { git init }
-        Exec { git config user.email "regression-test@example.com" }
-        Exec { git config user.name "regression-test" }
-        Exec { git add -A . }
-        Exec { git commit -m "Initial commit with expected files" | out-null }
-        Get-ChildItem * -Include *.json -Recurse | Remove-Item -Force
-    }
+    CopyFolderRecursively -from .\expected -to .\actual -exclude "*${ruleId}.json"
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/its/regression-test.ps1]: [minor] We try to respect the pascal case for arguments: `-From`, `-To` and `-Exclude`.

File: sonaranalyzer-dotnet/its/regression-test.ps1 [hunk 2]
```diff
@@ -25,33 +35,121 @@ function Build-Project([string]$ProjectName, [string]$SolutionRelativePath) {
         /fl /flp:"logFile=output\$ProjectName.log;verbosity=d"
 }
 
-function Initialize-ActualFolder {
+function Initialize-ActualFolder() {
+    $timer = [system.diagnostics.stopwatch]::StartNew()
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/its/regression-test.ps1]: Is it meant to stay there or was it only for test purposes?

File: sonaranalyzer-dotnet/its/regression-test.ps1 [hunk 3]
```diff
@@ -25,33 +35,121 @@ function Build-Project([string]$ProjectName, [string]$SolutionRelativePath) {
...
-        Exec { git commit -m "Initial commit with expected files" | out-null }
-        Get-ChildItem * -Include *.json -Recurse | Remove-Item -Force
-    }
+    CopyFolderRecursively -from .\expected -to .\actual -exclude "*${ruleId}.json"
+    Write-Host "Initialize-ActualFolder:" $timer.Elapsed.TotalSeconds
 }
 
-function Initialize-OutputFolder {
+function Initialize-OutputFolder() {
+    $timer = [system.diagnostics.stopwatch]::StartNew()
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/its/regression-test.ps1]: Is it meant to stay there or was it only for test purposes?

File: sonaranalyzer-dotnet/its/regression-test.ps1 [hunk 4]
```diff
@@ -25,33 +35,121 @@ function Build-Project([string]$ProjectName, [string]$SolutionRelativePath) {
...
+        Set-Content -Path ".\output\AllSonarAnalyzerRules.ruleset" -Value $rulesetWithOneRule
+    }
+    else {
+        Copy-Item ".\AllSonarAnalyzerRules.ruleset" -Destination ".\output"
+    }
+    Write-Host "Initialize-OutputFolder:" $timer.Elapsed.TotalSeconds
+}
+
+function Export-AnalyzerPerformancesFromLogs([string[]]$buildLogsPaths) {
+    return $buildLogsPaths |
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/its/regression-test.ps1]: [minor][cosmetic] I would put the piping operator on the next line for all calls, it makes things more readable IMO.

File: sonaranalyzer-dotnet/its/regression-test.ps1 [hunk 5]
```diff
@@ -25,33 +35,121 @@ function Build-Project([string]$ProjectName, [string]$SolutionRelativePath) {
...
+        Group-Object Rule |
+        Foreach-Object {
+            New-Object PSObject -property @{
+                Rule = $_.Name;
+                Time = [math]::Round(($_.Group.Time | Measure-Object -sum).Sum, 3)}
+        } |
+        Sort-Object Time -Descending
+}
+
+function GetFullPath($folder) {
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/its/regression-test.ps1]: We have tried to always comply to powershell naming guideline. Please rename this method to comply to it.

File: sonaranalyzer-dotnet/its/regression-test.ps1 [hunk 6]
```diff
@@ -25,33 +35,121 @@ function Build-Project([string]$ProjectName, [string]$SolutionRelativePath) {
...
+    return [System.IO.Path]::GetFullPath((Join-Path (Get-Location).Path $folder))
+}
+
+function CopyFolderRecursively($from, $to, $include, $exclude) {
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/its/regression-test.ps1]: We have tried to always comply to powershell naming guideline. Please rename this method to comply to it.

File: sonaranalyzer-dotnet/its/regression-test.ps1 [hunk 7]
```diff
@@ -25,33 +35,121 @@ function Build-Project([string]$ProjectName, [string]$SolutionRelativePath) {
...
+    $fromPath = GetFullPath -folder $from
+    $toPath = GetFullPath -folder $to
+
+    $files = if ($include)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/its/regression-test.ps1]: This if/else is hard to read. Please reformat it properly.

File: sonaranalyzer-dotnet/its/regression-test.ps1 [hunk 8]
```diff
@@ -25,33 +35,121 @@ function Build-Project([string]$ProjectName, [string]$SolutionRelativePath) {
...
+        Format-Table -Wrap -AutoSize | Out-String -Width 100
+    $timings | Where-Object { $_.Rule -match 'SonarAnalyzer\.VisualBasic.*' }           |
+        Format-Table -Wrap -AutoSize | Out-String -Width 100
+    $timings | Where-Object { $_.Rule -match 'SonarAnalyzer\.Rules\.CSharp.*' }         |
+        Format-Table -Wrap -AutoSize | Out-String -Width 100
+    $timings | Where-Object { $_.Rule -match 'SonarAnalyzer\.Rules\.VisualBasic.*' }    |
+        Format-Table -Wrap -AutoSize | Out-String -Width 100
+}
+
+function DiffResults() {
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/its/regression-test.ps1]: We have tried to always comply to powershell naming guideline. Please rename this method to comply to it.

File: sonaranalyzer-dotnet/its/regression-test.ps1 [hunk 9]
```diff
@@ -25,33 +35,121 @@ function Build-Project([string]$ProjectName, [string]$SolutionRelativePath) {
...
+
+    CopyFolderRecursively -from .\expected -to .\diff\expected -include "*${ruleId}.json"
+    CopyFolderRecursively -from .\actual   -to .\diff\actual -include "*${ruleId}.json"
+
+    Exec { & git diff --no-index --quiet --exit-code .\diff\expected .\diff\actual } `
+          -errorMessage $errorMsg
 }
 
 try {
+    $sw = [system.diagnostics.stopwatch]::StartNew()
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/its/regression-test.ps1]: Is it meant to stay there or was it only for test purposes?

File: sonaranalyzer-dotnet/its/regression-test.ps1 [hunk 10]
```diff
@@ -72,17 +170,19 @@ try {
     Write-Header "Processing analyzer results"
 
     Write-Host "Normalizing the SARIF reports"
-    Exec { & .\create-issue-reports.ps1 }
+    $timer2 = [system.diagnostics.stopwatch]::StartNew()
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/its/regression-test.ps1]: Is it meant to stay there or was it only for test purposes?

File: sonaranalyzer-dotnet/its/regression-test.ps1 [hunk 11]
```diff
@@ -72,17 +170,19 @@ try {
...
+        Exec { & .\create-issue-reports.ps1 }
+    Write-Host "Normalizing the SARIF reports:" $timer2.Elapsed.TotalSeconds
 
     Write-Host "Computing analyzer performance"
-    Exec { & .\extract-analyzer-performances.ps1 }
+    $timer3 = [system.diagnostics.stopwatch]::StartNew()
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/its/regression-test.ps1]: Is it meant to stay there or was it only for test purposes?

File: sonaranalyzer-dotnet/its/regression-test.ps1 [hunk 12]
```diff
@@ -72,17 +170,19 @@ try {
...
+    Measure-AnalyzerPerformance
+    Write-Host "Computing analyzer performance:" $timer3.Elapsed.TotalSeconds
 
     Write-Host "Checking for differences..."
-    Invoke-InLocation "actual" {
-        Exec { git add -A . }
-        Exec { git diff --cached --exit-code | out-null } -errorMessage `
-            "ERROR: There are differences between the actual and the expected issues."
-    }
+    $timer4 = [system.diagnostics.stopwatch]::StartNew()
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/its/regression-test.ps1]: Is it meant to stay there or was it only for test purposes?

File: sonaranalyzer-dotnet/its/regression-test.ps1 [hunk 13]
```diff
@@ -96,4 +196,12 @@ finally {
     Write-Host "Removing the import before target file from '${msBuildImportBefore}'"
     Remove-Item -Force (Join-Path $msBuildImportBefore "\SonarAnalyzer.Testing.ImportBefore.targets") `
         -ErrorAction Ignore
+
+    $sw.Stop()
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/its/regression-test.ps1]: Is it meant to stay there or was it only for test purposes?

File: sonaranalyzer-dotnet/its/update-expected.ps1 [hunk 14]
```diff
@@ -6,9 +6,16 @@ $ErrorActionPreference = "Stop"
 try {
     Push-Location $PSScriptRoot
 
+    if (-not (Test-Path .\actual)) {
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/its/update-expected.ps1]: Let's try to always use Pascal case.

File: sonaranalyzer-dotnet/its/regression-test.ps1 (hunk 13 above)
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/its/regression-test.ps1]: I prefer to keep all of them, so if we have any slowness in the future, the perf data is already there.

Comment (Evangelink) [reply sonaranalyzer-dotnet/its/regression-test.ps1]: Ok to keep them all but then start using a proper variable name. `sw`, `timer1` and so on don't help understanding what they measure.

File: sonaranalyzer-dotnet/its/regression-test.ps1 (hunk 8 above)
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/its/regression-test.ps1]: We (I mean in my past life) followed the guideline that only public/exported functions follow the Verb-Noun, and the private functions don't - so it's very easy to tell what's safe to change and what is potentially used anywhere. I'll follow that for new scripts, but will adjust this one as you suggest. Mixing two styles in one script is definitely bad.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/DiagnosticTagsHelper.cs [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+using Microsoft.CodeAnalysis;
+
+namespace SonarAnalyzer.Helpers
+{
+    public static class DiagnosticTagsHelper
+    {
+        public static string[] NotConfigurableTag()
+        {
+#if DEBUG
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/DiagnosticTagsHelper.cs]: A comment explaining why a different tag is returned for debug builds would be useful.

**Pull request merged**